Gems/Atom/RHI/Code/Source/RHI/MultiDevicePipelineState.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/RHI.Reflect/InputStreamLayout.h>
+#include <Atom/RHI.Reflect/PipelineLayoutDescriptor.h>
+#include <Atom/RHI/Factory.h>
+#include <Atom/RHI/MultiDevicePipelineLibrary.h>
+#include <Atom/RHI/MultiDevicePipelineState.h>
+#include <Atom/RHI/RHISystemInterface.h>
+
+namespace AZ::RHI
+{
+    bool MultiDevicePipelineState::ValidateNotInitialized() const
+    {
+        if (Validation::IsEnabled())
+        {
+            if (IsInitialized())
+            {
+                AZ_Error("MultiDevicePipelineState", false, "MultiDevicePipelineState already initialized!");
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    ResultCode MultiDevicePipelineState::Init(
+        MultiDevice::DeviceMask deviceMask, const PipelineStateDescriptor& descriptor, MultiDevicePipelineLibrary* pipelineLibrary)
+    {
+        if (!ValidateNotInitialized())
+        {
+            return ResultCode::InvalidOperation;
+        }
+
+        MultiDeviceObject::Init(deviceMask);
+
+        ResultCode resultCode = ResultCode::Success;
+
+        switch (descriptor.GetType())
+        {
```

Comment:
I think you can move the switch statement inside IterateDevices.  The switch statement should only encapsulate the m_devicePipelineStates[deviceIndex]->Init call

Reply:
I switched it around :+1: